File: elixir/lib/name_normalizer.ex
defmodule NameNormalizer do
  def normalize(name), do:
    name |> String.split(", ") |> normalize_name_and_maybe_suffix

  defp normalize_name_and_maybe_suffix([name, suffix]), do:
    (with {:ok, normalized_name} <- normalize_name_and_maybe_suffix([name]) do
      {:ok, "#{normalized_name}, #{suffix}"}
    end)

  defp normalize_name_and_maybe_suffix([name]), do:
    {:ok, name |> String.split |> handle_parts }
  defp normalize_name_and_maybe_suffix(_), do:
    {:error, "Too many commas"}

  defp handle_parts([]), do: ""
  defp handle_parts([single_name]), do: single_name
  defp handle_parts([first_name, last_name]), do: "#{last_name}, #{first_name}"
  defp handle_parts([first_name | other_names]) do
    {middle_names, last_name} = Enum.split(other_names, -1)
    Enum.join(["#{last_name},", first_name, initials(middle_names)], " ")
  end

  defp initials(middle_names), do:
    middle_names |> Enum.map(&initial/1) |> Enum.join(" ")

  defp initial(<<_>> = name), do: String.first(name)
  defp initial(name), do: "#{String.first(name)}."
end

File: python/name_normalizer.py
def normalize(name: str) -> str:
    trimmed = name.strip()

    (names, suffix) = trimmed.split(",") if "," in trimmed else (trimmed, None)
    result = build_name(names.split(" "))

    if suffix:
        return f"{result}, {suffix.strip()}"

    return result


def build_name(names) -> str:
    if len(names) == 0:
        return ""
    if len(names) == 1:
        return names[0]
    if len(names) == 2:
        return names[1] + ", " + names[0]
    if len(names) > 2:
        return f"{names[-1]}, {names[0]} {handle_middle_names(names[1:-1])}"


def handle_middle_names(middle_names) -> str:
    if len(middle_names) == 1 and len(middle_names[0]) == 1:
        return middle_names[0]

    return " ".join([name[0] + "." for name in middle_names if len(name) > 0])

